ironfish-rust/src/public_account/description.rs

```diff
@@ -0,0 +1,218 @@
+            transfers.push(transfer);
+        }
+        let address = PublicAddress::read(&mut reader)?;
+
+
+        Ok(Self {
+            version,
+            threshold,
+            signers,
+            signatures,
+            address,
+            transfers,
+        })
+    }
+
+    pub fn write<W: io::Write>(&self, mut writer: W) -> Result<(), IronfishError> {
+        // TODO: think about ideal ordering here
+        writer.write_all(&self.version.to_le_bytes())?;
+        
+        writer.write_all(&self.threshold.to_le_bytes())?;
+
+        let signers_len = self.signers.len() as i16;
+        writer.write_all(&signers_len.to_le_bytes())?;
+
+        for signer in &self.signers {
+            writer.write_all(signer.as_bytes())?;
+        }
+
+        let signatures_len = self.signatures.len() as i16;
+        writer.write_all(&signatures_len.to_le_bytes())?;
+
+        for signature in &self.signatures {
+            writer.write_all(&signature.to_bytes())?;
+        }
+
+        let transfers_len = self.transfers.len() as i16;
+        writer.write_all(&transfers_len.to_le_bytes())?;
+
+        for transfer in &self.transfers {
+            transfer.write(&mut writer)?;
+        }
+
+        self.address.write(&mut writer)?;
+
+        Ok(())
+    }
+
+    pub fn valid(&self) -> Result<(), IronfishError> {
+        if self.threshold < 1 {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        if self.signers.len() < self.threshold as usize {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidThreshold));
+        }
+
+        if self.signatures.len() < self.threshold as usize {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidData));
+        }
+
+        Ok(())
+    }
+
+    pub fn verify(&self) -> Result<(), IronfishError> {
+        self.valid()?;
+
+        let hash =
+            &PublicAccountDescription::hash(&self.version, &self.threshold, &self.signers, &self.address, &self.transfers)?;
+        let is_valid = self
+            .signers
+            .iter()
+            .zip(&self.signatures)
+            .any(|(signer, signature)| signer.verify(hash, signature).is_ok());
+        if !is_valid {
+            return Err(IronfishError::new(IronfishErrorKind::InvalidSignature));
+        }
+        Ok(())
+    }
+
+    pub fn hash(
+        version: &u8,
+        threshold: &i16,
+        signers: &Vec<VerifyingKey>,
+        address: &PublicAddress,
+        transfers: &Vec<Transfer>,
+    ) -> Result<[u8; 32], IronfishError> {
+        // TODO(jwp): verify which hashers supported by axelar
+        let mut hasher = blake3::Hasher::new();
+        hasher.update(&version.to_le_bytes());
+        hasher.update(&threshold.to_le_bytes());
+        for signer in signers {
+            hasher.update(signer.as_bytes());
+        }
+        for transfer in transfers {
+            hasher.update(&transfer.as_bytes()?);
+        }
```

Comment:
Variable-length items should be prefixed by their length. This makes the input to the hash "prefix free".

Not including a length prefix can potentially allow somebody to construct two distinct inputs that hash the same. For example, input 1 could be `signer 1 || signer 2 || transfer 1 || transfer 2`, and input 2 could be `singer 1 || signer 2 || signer 3 || transfer 2`, where `transfer 1` and `signer 3` have the same contents. The contents are semantically different, but would result in the same hash.

While this specific example might not result in a successful attack in this context, it's best to entirely close the door to this class of attacks.

Reply:
Great call!